samples-check/src/main/java/org/gradle/exemplar/test/engine/SamplesRunnerJUnitEngine.java

```diff
@@ -0,0 +1,313 @@
+    public static final List<String> SAFE_SYSTEM_PROPERTIES = Arrays.asList("file.separator", "java.home", "java.vendor", "java.version", "line.separator", "os.arch", "os.name", "os.version", "path.separator", "user.dir", "user.home", "user.name");
+
+    private final File tempDir;
+
+    public SamplesRunnerJUnitEngine() throws IOException {
+        tempDir = Files.createTempDirectory("samples-junit-engine").toFile();
+    }
+
+    @Override
+    public String getId() {
+        return "SamplesRunnerJUnitEngine";
+    }
+
+    @Override
+    public TestDescriptor discover(EngineDiscoveryRequest discoveryRequest, UniqueId uniqueId) {
+        EngineDescriptor engineDescriptor = new EngineDescriptor(uniqueId, "Samples Runner JUnit Engine");
+        discoveryRequest.getSelectorsByType(ClassSelector.class)
+                .stream()
+                .map(ClassSelector::getJavaClass).forEach(javaClass -> {
+                    filterClass(uniqueId, engineDescriptor, javaClass);
+                });
+
+        return engineDescriptor;
+    }
+
+    private void filterClass(UniqueId uniqueId, EngineDescriptor engineDescriptor, Class<?> javaClass) {
+        Samples samplesDef = javaClass.getAnnotation(Samples.class);
+        if (samplesDef != null) {
+            UniqueId classUniqueId = uniqueId.append("className", javaClass.getSimpleName());
+            SamplesTestDescriptor samplesTestDescriptor = new SamplesTestDescriptor(
+                    classUniqueId,
+                    javaClass
+            );
+
+            List<Sample> samples = samplesDef.samplesType() == Samples.SamplesType.DEFAULT ?
+                    SamplesDiscovery.externalSamples(getSamplesRootDir(
+                            samplesDef.root(),
+                            samplesDef.implicitRootDirSupplier()
+                    )) :
+                    SamplesDiscovery.embeddedSamples(getSamplesRootDir(
+                            samplesDef.root(),
+                            samplesDef.implicitRootDirSupplier()
+                    ));
+
+            List<? extends OutputNormalizer> normalizers = instantiateList(samplesDef.outputNormalizers());
+            List<SampleModifier> sampleModifiers = instantiateList(samplesDef.modifiers());
+            Function<CommandExecutorParams, CommandExecutor> commandExecutorFunction = instantiateObject(samplesDef.commandExecutorFunction());
+
+            samples.forEach(sample ->
+                    samplesTestDescriptor.addChild(new SampleTestDescriptor(
+                            classUniqueId.append("sampleId", sample.getId()),
+                            sample,
+                            commandExecutorFunction,
+                            normalizers,
+                            sampleModifiers
+                    ))
+            );
+
+            engineDescriptor.addChild(samplesTestDescriptor);
+        }
+    }
+
+    @Override
+    public void execute(ExecutionRequest request) {
```

Comment:
Extending `HierarchicalTestEngine` would allow to move the execution logic to the `TestDescriptor`.